org.eclipse.xtext.xbase/src/org/eclipse/xtext/xbase/compiler/XbaseCompiler.java

```diff
@@ -460,12 +461,55 @@ protected void _toJavaExpression(XBlockExpression expr, ITreeAppendable b) {
 		b.append(getVarName(expr, b));
 	}
 
-	protected void _toJavaStatement(XTryCatchFinallyExpression expr, ITreeAppendable outerAppendable, boolean isReferenced) {
+	protected void _toJavaStatement(XTryCatchFinallyExpression expr, ITreeAppendable outerAppendable,
+			boolean isReferenced) {
 		ITreeAppendable b = outerAppendable.trace(expr, false);
+		GeneratorConfig config = b.getGeneratorConfig();
+		boolean java7 = false;  // If Java 7 or better: use Java's try with resources
+		if (config != null && config.getJavaSourceVersion().isAtLeast(JAVA7))
+			java7 = true;
+		EList<XVariableDeclaration> resources = expr.getResources();
+		boolean isTryWithRes = !resources.isEmpty();
+
 		if (isReferenced && !isPrimitiveVoid(expr)) {
 			declareSyntheticVariable(expr, b);
 		}
-		b.newLine().append("try {").increaseIndentation();
+
+		// Resources    defined before try-statement, for java versions < 7
+		if (isTryWithRes && !java7) {
+			for (XVariableDeclaration res : resources) {
+				b.newLine();
+				LightweightTypeReference type = appendVariableTypeAndName(res, b);
+				b.append(" = ");
+				if (res.getRight() instanceof XConstructorCall)
+					constructorCallToJavaExpression((XConstructorCall) res.getRight(), b);
```

Comment:
The constructor invocation should happen in the try block. If the constructor defines an exception to be thrown, it can be caught by a catch of the try block.